app/services/structured_fact_importer.rb

```diff
@@ -19,6 +19,28 @@ def normalize_recurse(memo, facts, prefix = '')
     memo
   end
 
+  def find_or_create_fact_name(name, value)
+    if name.include?(FactName::SEPARATOR)
+      parent_name = /(.*)#{FactName::SEPARATOR}/.match(name)[1]
+      parent_fact = find_or_create_fact_name(parent_name, nil)
+      fact_name = parent_fact.children.where(:name => name, :type => fact_name_class.to_s).first
+    else
+      parent_fact = nil
+      fact_name = fact_name_class.where(:name => name, :type => fact_name_class.to_s).first
+    end
+
+    if fact_name
+      fact_name.update_attribute(:compose, value.nil?) if value.nil? && !fact_name.compose?
+    else
+      fact_name = fact_name_class.create!(:name => name, :type => fact_name_class.to_s, :compose => value.nil?, :parent => parent_fact)
+    end
+    fact_name
+  end
+
+  def create_fact_name(fact_names, name, fact_value)
+    fact_names[name] ||= fact_name_class.create!(:name => name)
+  end
+
   def create_fact_name(fact_names, name, fact_value)
```

Comment:
Oh these methods are not necessary anymore. I will remove that in the next push, they are unused.